### PR TITLE
Adding support for `Function version`

### PR DESCRIPTION
**Pull request by Vandita2020**
**Summary**:
Adding support to publish `Version` for a `Function`.

**Description**:
This PR adds implementation of AWS Lambda’s `Function Version` feature in the ACK. Function versions allow developers to publish new versions of their Lambda functions while maintaining older versions for backward compatibility or testing purposes. This will allow developers to publish new versions of their lambda functions and manage function updates seamlessly using ACK.

`Version` resource for a Function can be created using following manifest:
```
apiVersion: lambda.services.k8s.aws/v1alpha1
kind: Version
metadata:
  name: $VERSION_NAME
  annotations:
    services.k8s.aws/region: $AWS_REGION
spec:
  functionName: $FUNCTION_NAME
  codeSHA256: $HASH
  revisionID: $REVISION_ID
  description: version created by ACK lambda-controller e2e tests
```

The PR supports creation and deletion of `Version` resource. It also includes E2E tests to verify the functionality.

**Acknowledgment**:
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by a-hilaly — COMMENTED** [review]
Since we already have a `LayerVersion` resource, and to avoid UX confusions, should we name this resource `FunctionVersion` instead?

**Comment by valerena** [comment]
> Since we already have a `LayerVersion` resource, and to avoid UX confusions, should we name this resource `FunctionVersion` instead?

It's a good call, but the resource is called just `Version` in other places, like the API is just called [`PublishVersion`](https://docs.aws.amazon.com/lambda/latest/dg/API_PublishVersion.html), and there's an [`AWS::Lambda::Version`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-version.html) resource in CFN and a [`Version`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_lambda.Version.html) construct in CDK. 

I don't know what would create more confusion really

File: generator.yaml [hunk 1]
```diff
@@ -7,11 +7,28 @@ ignore:
     # FunctionUrlConfig
     # LayerVersion
 operations:
-  PublishLayerVersion:
+  PublishLayerVersion: #opID
```
Comment (valerena) [inline generator.yaml:10]: Is this comment needed or relevant? (can we remove it?)

File: templates/hooks/version/sdk_create_pre_build_request.go.tpl [hunk 2]
```diff
@@ -0,0 +1,18 @@
+res := &svcsdk.ListVersionsByFunctionInput{}
+res.FunctionName = desired.ko.Spec.FunctionName
+var list *svcsdk.ListVersionsByFunctionOutput
+list, err = rm.sdkapi.ListVersionsByFunctionWithContext(ctx, res)
+if err != nil {
+	return nil, err
+}
+bigList := list.Versions
```
Comment (valerena) [inline templates/hooks/version/sdk_create_pre_build_request.go.tpl:8]: Maybe `bigList` should be called something more specific, like `versionsList` or `previousVersions` (Or something that makes more sense, for when it's used in the `sdk_create_pre_set_output.go.tpl` file)

File: generator.yaml (hunk 1 above)
Comment (Vandita2020) [reply generator.yaml]: My mistake, have removed it now. Thanks!

File: templates/hooks/version/sdk_create_pre_build_request.go.tpl (hunk 2 above)
Comment (Vandita2020) [reply templates/hooks/version/sdk_create_pre_build_request.go.tpl]: Makes sense, I went with `versionList`. We're getting list of Versions for `LayerVersion` resource and assigned it to variable named `versionList` only, made sense to use the same naming for this case too. Let me know if we should keep the naming different.

File: pkg/resource/version/sdk.go [hunk 3]
```diff
@@ -0,0 +1,961 @@
...
+	versionList := list.Versions
+
+	for ok := list.NextMarker != nil; ok; ok = (list.NextMarker != nil) {
+		res.Marker = list.NextMarker
+		list, err = rm.sdkapi.ListVersionsByFunctionWithContext(ctx, res)
+		if err != nil {
+			return nil, err
+		}
+		versionList = append(versionList, list.Versions...)
+	}
```
Comment (a-hilaly) [inline pkg/resource/version/sdk.go:457]:
```suggestion
    marker := nil
	
	versionsList := []svcsdk.Version
	for {
		listVersionsInput := &svcsdk.ListVersionsByFunctionInput{
	        FunctionName: desired.ko.Spec.FunctionName,
	        Marker: marker
	    }
	    listVersionResponse, err := rm.sdkapi.ListVersionsByFunctionWithContext(ctx, listVersionsInput)
	    if err != nil {
		    return nil, err
	    }
		versionsList = append(versionLists, listVersionResponse.Versions...)
		
		if listVersionResponse.NextMarker == nil {
		    break
		}
		marker = listVersionResponse.NextMarker 
	}
```

File: test/e2e/resources/version_with_check.yaml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+apiVersion: lambda.services.k8s.aws/v1alpha1
+kind: Version
+metadata:
+  name: $VERSION_NAME
+  annotations:
+    services.k8s.aws/region: $AWS_REGION
+spec:
+  functionName: $FUNCTION_NAME
+  codeSHA256: $HASH
```
Comment (a-hilaly) [inline test/e2e/resources/version_with_check.yaml:9]: How is the `$HASH` set

**Review by a-hilaly — COMMENTED** [review]
Ty @Vandita2020 - left two comments inline.

File: test/e2e/resources/version_with_check.yaml (hunk 4 above)
Comment (Vandita2020) [reply test/e2e/resources/version_with_check.yaml:9]: `codeSHA256` and `RevisionId` fields are optional and are used to check whether the Version is published only when there is some change in the code or its configuration. I have added new E2E test to verify the code by setting $HASH to publish the Version.

File: pkg/resource/version/sdk.go (hunk 3 above)
Comment (Vandita2020) [reply pkg/resource/version/sdk.go]: Made the suggested changes. Thanks!

File: test/e2e/resources/functionVersion.yaml
Comment (a-hilaly) [inline test/e2e/resources/functionVersion.yaml:1]: Can we name this file `function_version.yaml` , to keep the same style used for the files in the same directory?

File: test/e2e/tests/test_version.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        # Delete k8s resource
+        _, deleted = k8s.delete_custom_resource(ref)
+        assert deleted is True
+
+        time.sleep(DELETE_WAIT_AFTER_SECONDS)
+
+        # Check function version doesn't exist
+        assert not lambda_validator.version_exists(lambda_function_name, version_number)
+
+    def test_version_with_check(self, lambda_client, lambda_function):
```
Comment (a-hilaly) [inline test/e2e/tests/test_version.py:132]: What does check here means?

File: apis/v1alpha1/version.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+	// is a string array with one of the valid values. The default architecture
+	// value is x86_64.
+	// +kubebuilder:validation:Optional
+	Architectures []*string `json:"architectures,omitempty"`
+	// The size of the function's deployment package, in bytes.
+	// +kubebuilder:validation:Optional
+	CodeSize *int64 `json:"codeSize,omitempty"`
+	// The function's dead letter queue.
+	// +kubebuilder:validation:Optional
+	DeadLetterConfig *DeadLetterConfig `json:"deadLetterConfig,omitempty"`
```
Comment (a-hilaly) [inline apis/v1alpha1/version.go:79]: Most of the fields in here are not really relevant to a function, should we get rid of them?

File: test/e2e/resources/functionVersion.yaml
Comment (Vandita2020) [reply test/e2e/resources/functionVersion.yaml:1]: This file is similar to `version_with_check.yaml` file, so I deleted it. Don't remember how it got here.

File: test/e2e/tests/test_version.py (hunk 5 above)
Comment (Vandita2020) [reply test/e2e/tests/test_version.py:132]: To check that the version we created above got deleted. We call `getFunction` API and pass the Version number along with Function name as we are only deleting a version and not the whole function.

File: apis/v1alpha1/version.go (hunk 6 above)
Comment (Vandita2020) [reply apis/v1alpha1/version.go:79]: We're using [`PublishVersion`](https://docs.aws.amazon.com/lambda/latest/dg/API_PublishVersion.html) API to set the status and its response has all these fields. So we cannot do anything with the extra fields.

**Pull request merged**